### PR TITLE
[ workflows ] use YAML references to prune duplicate file lists

### PR DESCRIPTION
**Pull request by andreasabel**
[ workflows ] use YAML references to prune duplicate file lists

Suggested by https://github.com/agda/agda/issues/5553#issuecomment-918264936

### CODE REVIEW

**Comment by andreasabel** [comment]
Unfortunately, GHA doesn't support anchors in yaml files.
```
The workflow is not valid. .github/workflows/haddock.yml: Anchors are not currently supported. Remove the anchor 'trigger_path_list'
```
We could of course expand the anchors by converting to JSON and back: `yaml2json test.yml | json2yaml -`
Once we generate the `.yml` files, we can use more abbreviation mechanism that are outside of YAML...

**Comment by L-TChen** [comment]
Or, we can use Dhall to generate YAML?

**Comment by L-TChen** [comment]
It is odd that my `make` complains 
```
make: *** No rule to make target ...
```
for every YAML configuration that I want to generate. Which `make` do you use?

**Comment by andreasabel** [comment]
@L-TChen : It is `gmake`:
```
$ make --version
GNU Make 4.2.1
Built for x86_64-apple-darwin18.0.0
```

Could be there are other expansion rules for variables in patterns?  Or functions like `wildcard`, `notdir`, `sort` failing?
Does `make debug` print something meaningful?

**Comment by L-TChen** [comment]
I tried `gmake` 3.81 (shipped with macOS) and 4.3 (installed via `homebrew`), but both of them cannot work. 

`gmake debug` prints the following
```
srcpath = ../../src/github/workflows
targets = cabal-test.yml cabal.yml deploy.yml haddock.yml lint.yaml stack.yml test.yml user_manual.yml whitespace.yml
```
I assume it is the expected result, right?

**Comment by L-TChen** [comment]
Aha, @andreasabel you didn't commit `header.txt`!!

**Comment by andreasabel** [comment]
> Aha, @andreasabel you didn't commit `header.txt`!!

Mea culpa, very sorry to waste your time, @L-TChen.  (blush)

We should have a test (at least in the master `Makefile`) that the generation of the workflows works!

**Comment by andreasabel** [comment]
@L-TChen : I pushed `header.txt`, would you please try again?

**Comment by L-TChen** [comment]
> Mea culpa, very sorry to waste your time, @L-TChen. (blush)

No worries. It is a way for me to have a break from research.

It looks good to me now, but just remember to resolve conflicts before merging.

File: .github/workflows/Makefile [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+requirements: req-cabal
+
+# The flag no-exe is by default on, we need to turn it off.
+# https://github.com/snoyberg/yaml/issues/194
+
+req-cabal:
+	cabal install yaml -f-no-exe
+
+STACK_OPTS = --resolver nightly --system-ghc
```
Comment (L-TChen) [inline .github/workflows/Makefile]: Why is nightly LTS used?

**Review by L-TChen — APPROVED** [review]
Just remember to resolve conflicts and perhaps also remove the option `--resolver nightly` which seems unnecessary.

File: .github/workflows/Makefile (hunk 1 above)
Comment (andreasabel) [reply .github/workflows/Makefile]: Dunno.  I think because I am on GHC 9.0.1 which is only supported by `nightly`.

Removed it.

**Comment by andreasabel** [comment]
I added a `make workflows` target to the root `Makefile`.

@L-TChen Can you please whether your latest changes didn't get lost by conflict resolution?
I was a bit surprised that `git rebase master` just went through without reporting conflicts.

**Pull request merged**